src/lib.rs

```diff
@@ -0,0 +1,238 @@
+// Copyright (c) 2019 King's College London created by the Software Development Team
+// <http://soft-dev.org/>
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0>, or the MIT license <LICENSE-MIT or
+// http://opensource.org/licenses/MIT>, or the UPL-1.0 license <http://opensource.org/licenses/UPL>
+// at your option. This file may not be copied, modified, or distributed except according to those
+// terms.
+
+#![feature(alloc_layout_extra)]
+
+use std::alloc::{alloc, Layout};
+use std::mem::{size_of, transmute};
+use std::time::Instant;
+
+const REPS: usize = 100;
+const VEC_SIZE: usize = 10000000;
+const VAL_NOREAD: usize = 0xdeadbeef;
+const VAL_WITHREAD: usize = 0xFEEDC0DE;
+
+#[inline(never)]
+fn time<F>(mut f: F)
+where
+    F: FnMut(),
+{
+    let before = Instant::now();
+    for _ in 0..REPS {
+        f();
+    }
+    let d = Instant::now() - before;
+    println!("{:?}", d.as_secs() as f64 + d.subsec_nanos() as f64 * 1e-9);
+}
+
+// The trait whose function we'll dynamically look up.
+trait T {
+    fn m(&self) -> usize;
+}
+
+// This is a simple helper trait that lets us cut a lot of our code in half below: it allows us to
+// instantiate SNoRead or SWithRead via a trait. We can't put this in T because T then wouldn't be
+// a valid trait object.
+trait N {
+    fn new() -> Self;
+}
+
+// SNoRead and SWithRead are the same size, but SNoRead.m() doesn't actually read the "_i"
+// attribute (hence the leading underscore) whereas SWithRead does read from its "i" attribute
+// (hence the lack of a leading underscore).
+struct SNoRead {
+    _i: usize,
+}
+
+impl T for SNoRead {
+    fn m(&self) -> usize {
+        VAL_NOREAD
+    }
+}
+
+impl N for SNoRead {
+    fn new() -> Self {
+        SNoRead { _i: VAL_NOREAD }
+    }
+}
+
+struct SWithRead {
+    i: usize,
+}
+
+impl T for SWithRead {
+    fn m(&self) -> usize {
+        self.i
+    }
+}
+
+impl N for SWithRead {
+    fn new() -> Self {
+        SWithRead { i: VAL_WITHREAD }
+    }
+}
+
+fn vec_normal<S: 'static + N + T>() -> Vec<Box<dyn T>> {
+    let mut v = Vec::<Box<dyn T>>::with_capacity(VEC_SIZE);
+    for _ in 0..VEC_SIZE {
+        v.push(Box::new(S::new()));
```

Comment:
Did you push the elements one-by-one on purpose? If so, why?

(Not saying it's bad, just curious).

Reply:
I don't think there's another way to do this, because in this case `S` isn't cloneable.

Reply:
I see. No worries.